src/Server/Completion/TokenContainer.php

```diff
@@ -0,0 +1,80 @@
+<?php
+namespace LanguageServer\Server\Completion;
+
+use LanguageServer\Protocol\Position;
+use LanguageServer\Protocol\Range;
+
+class TokenContainer
+{
+
+    const T_SINGLE_CHAR = 10000;
+
+    /**
+     *
+     * @var \LanguageServer\Server\Completion\Token[]
+     */
+    private $tokens;
+
+    public function __construct(array $tokens)
```

Comment:
This simple example will return null instead of ast:
```
<?php
function abc(){
}
function test(){
    $
```
From my personal experience I'm concerned that there are too much cases to cover to have very good parser error recovery. I will explain more about tokens in next comment.